src/D_functions_and_procedures.sql

```diff
@@ -23,6 +23,44 @@ LANGUAGE 'plpgsql';
 --D2 TEST
 SELECT * FROM simple_search('apple',1);
 
+-- Structured string search
+CREATE OR REPLACE FUNCTION structured_string_search (
+  p_title VARCHAR(100), 
+  p_plot VARCHAR(100), 
+  p_character VARCHAR(100), 
+  p_person VARCHAR(100), 
+  p_user_id INTEGER
+)
+RETURNS TABLE (media_id INTEGER, title TEXT)
+AS $$
+BEGIN
+  -- SEARCH HISTORY
+  INSERT INTO search_history (p_user_id, type, query)
+  VALUES (user_id, 'structured_string_search', 
+          FORMAT('title: %s, plot: %s, character: %s, person: %s', p_title, p_plot, p_character, p_person));
+
+  -- RESULT
+  RETURN QUERY
+  WITH
+    search_result AS (
+        SELECT DISTINCT m.media_id, m.imdb_id, m."type" AS media_type
+        FROM media AS m
+        JOIN "release" r USING (media_id)
```

Comment:
You're doing a double `JOIN`, once in the `WITH` statement and once in the `SELECT` statement, I don't think this is necessary. _ #see line 59 as well._